torch/csrc/jit/codegen/cuda/kernel_resource_strings.h

```diff
@@ -257,7 +254,315 @@ void blockReduce(T& out, const T inp_val, Func reduction_op) {
+    offset += thread_idx.z;
+  if (Y_THREAD)
+    offset = offset * block_dim.y + thread_idx.y;
+  if (X_THREAD)
+    offset = offset * block_dim.x + thread_idx.x;
+  return offset;
+}
+
+/** Reduces all the reduction blocks in each reduction segment.
+
+  This is only used by one thread block per reduction segment. The input
+  reduction blocks of the segment are stored in an intermediate buffer pointed
+  by parameter in. Template parameters X/Y/Z_THREAD denote how the reduction
+  block is formed.
+
+  The size of a reduction block is by definition smaller or equal to the size of
+  a thread block. We use the remaining threads to parallelize reductions across
+  reduction blocks. For example, when X/Y/Z_THREAD = {true, false, false}, we
+  use blockDim.y*blockDim.z threads for each output value. This is done first by
+  loading the input values in parallel and then by reducing across threads of
+  dimensions whose XYZ_THREAD are false.
+
+  Note that what is done here after the loading from global memory is similar to
+  what the existing blockReduce function does. The main difference is that the
+  logical block to reduce is a 2D domain where the leading dimension is the size
+  of a reduction block and the second dimension is the remaining factor in each
+  thread block. For example, when X/Y/Z_THREAD = {false, true, false}, the
+  threads are arranged as (blockDim.y, blockDim.x*blockDim.z). We do not reduce
+  along the first dimension but only the second dimension. So, it is possible to
+  reuse the existing blockReduce with dim3{blockDim.y, blockDim.x*blockDim.z}
+  instead of blockDim and with X_THREAD and Y_THREAD being false and true,
+  respectively. Also, it still need to shuffle the final output values to their
+  actual corresponding threads. In the case of when X/Y/Z_THREAD = {false, true,
+  false}, after the intra-block reduction, the final results will still be held
+  by the first blockDim.y threads, which need to be transferred to threads at
+  threadIdx.x == 0 and threadIdx.z == 0.
+*/
+template <bool X_THREAD, bool Y_THREAD, bool Z_THREAD,
+          typename T, typename Func>
+__device__ void gridReduceLastBlock(T& out, const T *in, const size_t in_size,
+                                    Func reduction_op, T* shared_buf) {
+  const int tid = ioffset(threadIdx, blockDim);
+  const int block_size = isize(blockDim);
+  const int rblock_size = size_of_reduction_block<X_THREAD, Y_THREAD, Z_THREAD>(blockDim);
+
+  T inp = 0;
+  if (tid < in_size) {
+    inp = in[tid];
+  }
+  for (size_t i = tid + block_size; i < in_size; i += block_size) {
+    reduction_op(inp, in[i]);
+  }
+
+  const auto should_write = (X_THREAD || threadIdx.x == 0) &&
+      (Y_THREAD || threadIdx.y == 0) &&
+      (Z_THREAD || threadIdx.z == 0);
+
+  auto rem_size = block_size / rblock_size;
+
+  if (rem_size > 1) {
+    const int rblock_offset = tid % rblock_size;
+    const int rblock_idx = tid / rblock_size;
+    blockReduce<false, true, false>(
```

Comment:
This may be my naive understanding, but it seems to me that we're assuming `block_size` is evenly divisible by `rblock_size` is that the case?

Reply:
Yes, that is correct. The block size is the size of thread blocks. The reduction block is basically a sub domain of a thread block. Its share is defined as `{THREAD_X ? blockDim.x : 1, THREAD_Y ? blockDim.y : 1, THREAD_Z ? blockDim.z : 1}`. So, the block size is by definition always divisible by the reduction block size.